### PR TITLE
Publish diagnostics per file

### LINKED ISSUE #80: Diagnostics should be published per file
Right now they are published all together with the workspace URI.

### PR DESCRIPTION
**Pull request by natgabb**
Closes #80

### CODE REVIEW

File: groovy-language-server/src/main/java/com/palantir/ls/groovy/GroovyLanguageServerConfig.java [hunk 1]
```diff
@@ -70,13 +81,29 @@ public void setTelemetryEvent(Consumer<Object> telemetryEvent) {
...
-        this.publishDiagnostics = callback;
+    public void publishDiagnostics(Set<Diagnostic> diagnostics) {
+        if (diagnostics.isEmpty()) {
+            return;
+        }
+        Map<String, PublishDiagnosticsParamsBuilder> diagnosticsByFile = Maps.newHashMap();
+
+        diagnostics.forEach(diagnostic -> {
+            try {
+                String uri = Uris.absolutePathAsFileUri(Paths.get(diagnostic.getSource()));
```
Comment (jiahuijiang) [inline groovy-language-server/src/main/java/com/palantir/ls/groovy/GroovyLanguageServerConfig.java]: I think we can assume the diagnostic source is always a string in valid uri format

File: groovy-language-server/src/main/java/com/palantir/ls/groovy/GroovyLanguageServerConfig.java [hunk 2]
```diff
@@ -70,13 +81,29 @@ public void setTelemetryEvent(Consumer<Object> telemetryEvent) {
...
+                diagnosticsByFile.computeIfAbsent(uri, (value) -> new PublishDiagnosticsParamsBuilder().uri(uri))
+                        .diagnostic(diagnostic);
+            } catch (IllegalArgumentException e) {
+                logger.warn("Diagnostic had invalid URI source: {}", diagnostic);
+            }
+        });
+
+        diagnosticsByFile.values().stream().map(paramsBuilder -> paramsBuilder.build())
```
Comment (jiahuijiang) [inline groovy-language-server/src/main/java/com/palantir/ls/groovy/GroovyLanguageServerConfig.java:107]: diagnosticsByFile.values().stream().foreach(paramsBuilder -> publishDiagnostics.accept(paramsBuilder.build()))?

either way is fine :)

File: groovy-language-server/src/main/java/com/palantir/ls/groovy/GroovyLanguageServerConfig.java (hunk 1 above)
Comment (natgabb) [reply groovy-language-server/src/main/java/com/palantir/ls/groovy/GroovyLanguageServerConfig.java]: It's actually possible for the compiler to create errors with no source. We then default to "groovyc". Changing this to be nicer would be a really big refactor job.

Comment (natgabb) [reply groovy-language-server/src/main/java/com/palantir/ls/groovy/GroovyLanguageServerConfig.java]: Although I just realized we still want to send those diagnostics. But under which uri?

**Comment by natgabb** [comment]
@JiahuiJiang any more comments?

File: groovy-language-server/src/main/java/com/palantir/ls/groovy/GroovyLanguageServerConfig.java [hunk 3]
```diff
@@ -70,13 +76,33 @@ public void setTelemetryEvent(Consumer<Object> telemetryEvent) {
...
     }
 
     @Override
-    public void setPublishDiagnostics(Consumer<PublishDiagnosticsParams> callback) {
-        this.publishDiagnostics = callback;
+    public void publishDiagnostics(String workspaceUri, Set<Diagnostic> diagnostics) {
+        if (diagnostics.isEmpty()) {
+            return;
+        }
+        Map<String, PublishDiagnosticsParamsBuilder> diagnosticsByFile = Maps.newHashMap();
```
Comment (jiahuijiang) [inline groovy-language-server/src/main/java/com/palantir/ls/groovy/GroovyLanguageServerConfig.java]: Can make the key to be a Uri to be consistent with the reset of the codebase

File: groovy-language-server/src/main/java/com/palantir/ls/groovy/GroovyLanguageServerConfig.java (hunk 1 above)
Comment (natgabb) [reply groovy-language-server/src/main/java/com/palantir/ls/groovy/GroovyLanguageServerConfig.java]: fixed to use workspace uri in this case

**Comment by jiahuijiang** [comment]
+1 after last comment/suggestion

File: groovy-language-server/src/main/java/com/palantir/ls/groovy/GroovyLanguageServerConfig.java (hunk 3 above)
Comment (natgabb) [reply groovy-language-server/src/main/java/com/palantir/ls/groovy/GroovyLanguageServerConfig.java]: done

**Comment by jiahuijiang** [comment]
+1 after checks pass

**Pull request merged**